Expand common contractions and abbreviations

akhilaananthram opened the pull request:
Reads in a csv file containing contractions and their expansions then does the expansion.  Edit contractions.csv to add more contractions.

Also reads in a csv file containing domain specific abbreviations and expands them

Fixes #76 and #77

akhilaananthram commented:
@BoltzmannBrain

BoltzmannBrain commented:
@akhilaananthram please push some unit/integration tests go with this code. We don't need full coverage, but it'll be helpful to validate the text preprocessing works as expected as we modify it for experiments.

akhilaananthram commented:
@BoltzmannBrain: Do you want tests just for the new features to make sure it hasn't changed the basic functionality, or do you want tests for the old features as well since those don't exist right now?

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,169 @@
+
+    for e, p in zip(expected, predicted):
+      self.assertEqual(e, p, "Expected {} in list but got {}".format(e, p))
+
```
Two lines between methods.

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,169 @@
+    with self.assertRaises(KeyError) as context:
+      processor = TextPreprocess(contrTxt="abbreviations.csv")
+
+    self.assertIsInstance(context.exception, KeyError)
```
Don't need this.

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,169 @@
+    Tests that tokenizing behaves like no preprocessing if no contraction file
+    was provided
+    """
+    return
```
???

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,169 @@
+    expected_tokens = ["i", "can", "t", "work", "at", "identifier", "deleted",
+      "if", "you", "don", "t", "allw", "me", "to", "work", "from", "home"]
+    tokens = processor.tokenize(text, expandAbbr=True)
+    print tokens
```
Remove this

BoltzmannBrain commented:
@akhilaananthram the abbreviations and contractions functionality should be in their own methods, and definitely not all in the class init. A `TextPreprocess` instance shouldn't be required to process abbreviations and contractions. Please break these up logically into separate methods. And please edit the tests accordingly.

akhilaananthram commented:
So you want me to process abbreviations.csv and contractions.csv in their own methods?

BoltzmannBrain commented:
Yes, perhaps even more than one for each.

akhilaananthram commented:
@BoltzmannBrain: I pulled the process of reading in a csv into its own method and added a unit test for it

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,215 @@
+      expectedValue = expected[key]
+      actualValue = actual[key]
+      self.assertEqual(expectedValue, actualValue,
+        "Expected value {} for {}, but got {}".format(expectedValue, key, actualValue))
```
Wrap lines at 80 char.

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,215 @@
+    self.assertItemsEqual(actualKeys, expectedKeys)
+
+    for key in expectedKeys:
+      expectedValue = expected[key]
```
assigning variables like this is unnecessary, better to just assert `expected[key]` equal to `actual[key]`

BoltzmannBrain commented on tests/unit/utils/text_preprocess_test.py:
```diff
@@ -0,0 +1,215 @@
+    processor = TextPreprocess()
+    suffixes = ["", "s", "'s"]
+    abbreviations = processor.readExpansionFile("abbreviations.csv", "Abbr",
+        suffixes)
```
fix indent please

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -49,13 +53,60 @@ def __init__(self,
 
+    # Read in abbreviations
+    self.abbrs = self.readExpansionFile(abbrTxt, "Abbr", ["", "s", "'s"])
+    self.abbrRegex = re.compile(r"\b%s\b" % r"\b|\b".join(self.abbrs.keys()))
```
Difficult to tell what you're doing here; please add a (concise) comment.

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -49,13 +53,60 @@ def __init__(self,
+
+    # Read in contractions
+    self.contrs = self.readExpansionFile(contrTxt, "Contr")
+    # Don't want \b at the start of the word so 'll works
```
Please change this comment to be more clear what's going on here.

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -36,7 +38,9 @@ class TextPreprocess(object):
+               abbrTxt='abbreviations.csv',
+               contrTxt='contractions.csv'):
 
     # compilation.txt is a compilation of most frequent words from both British
```
Please replace this comment with a formal docstring for the __init__() method, where you briefly define what these three files are, and then also the hard-coded options in the methods -- i.e. expansion suffixes.

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -49,13 +53,60 @@ def __init__(self,
+          expSuffix = "{}{}".format(expansion, suffix)
+          expansionPairs[originalSuffix] = expSuffix
+    except IOError:
+      # To be backwards compatible, filename is not mandatory
```
What do you mean here?

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -49,13 +53,60 @@ def __init__(self,
+    @param suffixes        (list)      strings that are added to the end of
+                                       the original and expanded form if
+                                       provided
+    @return                (dict)      Maps original to expanded form
```
The return in the docstring should be explicit; e.g. "the keys are ___ with values for __."

BoltzmannBrain commented:
@akhilaananthram looks better, I left some more comments.

akhilaananthram commented:
@BoltzmannBrain: Please let me know if the additional comments are enough

BoltzmannBrain commented on fluent/utils/text_preprocess.py:
```diff
@@ -49,13 +68,71 @@ def __init__(self,
     self.txtCorpus = self.tokenize(self.corpus)
     self.bagOfWords = Counter(self.txtCorpus)
 
+  def readExpansionFile(self, filename, originalType, suffixes=None,
```
better (i.e. cleaner) to avoid passing flags like `originalType` whenever you can... pandas DataFrames have some useful functionalities that make this flag unnecessary

akhilaananthram commented:
@BoltzmannBrain: I removed the originalType flag and updated the tests for that function. So we want to allow the user to have any header format for the file? The reason I did that in the first place was because the contraction and abbreviation regular expressions are different. The abbreviation regular expression requires the entire word to match, while the contraction one only requires the end to match.

BoltzmannBrain commented:
Changes incorporated into #92.
